.github/workflows/unit-tests.yml

```diff
@@ -0,0 +1,56 @@
+            "openedx/core/djangoapps/ace_common/ openedx/core/djangoapps/agreements/ openedx/core/djangoapps/api_admin/ openedx/core/djangoapps/auth_exchange/ openedx/core/djangoapps/bookmarks/ openedx/core/djangoapps/cache_toolbox/ openedx/core/djangoapps/catalog/ openedx/core/djangoapps/ccxcon/ openedx/core/djangoapps/commerce/ openedx/core/djangoapps/common_initialization/ openedx/core/djangoapps/common_views/ openedx/core/djangoapps/config_model_utils/ openedx/core/djangoapps/content/ openedx/core/djangoapps/content_libraries/ openedx/core/djangoapps/contentserver/ openedx/core/djangoapps/cors_csrf/ openedx/core/djangoapps/course_date_signals/ openedx/core/djangoapps/course_groups/ openedx/core/djangoapps/coursegraph/ openedx/core/djangoapps/courseware_api/ openedx/core/djangoapps/crawlers/ openedx/core/djangoapps/credentials/ openedx/core/djangoapps/credit/ openedx/core/djangoapps/dark_lang/ openedx/core/djangoapps/debug/ openedx/core/djangoapps/demographics/ openedx/core/djangoapps/discussions/ openedx/core/djangoapps/django_comment_common/ openedx/core/djangoapps/embargo/ openedx/core/djangoapps/enrollments/ openedx/core/djangoapps/external_user_ids/",
+            "openedx/core/djangoapps/geoinfo/ openedx/core/djangoapps/header_control/ openedx/core/djangoapps/heartbeat/ openedx/core/djangoapps/lang_pref/ openedx/core/djangoapps/models/ openedx/core/djangoapps/monkey_patch/ openedx/core/djangoapps/oauth_dispatch/ openedx/core/djangoapps/olx_rest_api/ openedx/core/djangoapps/password_policy/ openedx/core/djangoapps/plugin_api/ openedx/core/djangoapps/plugins/ openedx/core/djangoapps/profile_images/ openedx/core/djangoapps/programs/ openedx/core/djangoapps/safe_sessions/ openedx/core/djangoapps/schedules/ openedx/core/djangoapps/self_paced/ openedx/core/djangoapps/service_status/ openedx/core/djangoapps/session_inactivity_timeout/ openedx/core/djangoapps/signals/ openedx/core/djangoapps/site_configuration/ openedx/core/djangoapps/system_wide_roles/ openedx/core/djangoapps/theming/ openedx/core/djangoapps/user_api/ openedx/core/djangoapps/user_authn/ openedx/core/djangoapps/util/ openedx/core/djangoapps/verified_track_content/ openedx/core/djangoapps/video_config/ openedx/core/djangoapps/video_pipeline/ openedx/core/djangoapps/waffle_utils/ openedx/core/djangoapps/xblock/ openedx/core/djangoapps/xmodule_django/ openedx/core/djangoapps/zendesk_proxy/ openedx/core/djangolib/ openedx/core/lib/ openedx/core/tests/ openedx/core/tests/ openedx/features/ openedx/testing/ openedx/tests/",
+            "cms/djangoapps/api/ cms/djangoapps/cms_user_tasks/ cms/djangoapps/course_creators/ cms/djangoapps/export_course_metadata/ cms/djangoapps/maintenance/ cms/djangoapps/models/ cms/djangoapps/pipeline_js/ cms/djangoapps/xblock_config/ cms/envs/ cms/lib/",
+            "cms/djangoapps/contentstore/",
+            "common/djangoapps/",
+            "common/lib/",
+        ]
```

Comment:
The ticket discusses the desire to avoid manual sharding because it leads to test interaction debt when the load is rebalanced, but it looks like that's what was done here. Is there a plan to dynamically generate the sharding?

If we're using manual sharing, there should also be documentation on constraints such as that all paths in a list entry must be in the same top-level directory (or whatever the constraint actually is), as well as documentation on how the list is to be re-generated.

Reply:
I think I also have a workable idea for automatic sharding. I'm not sure if this has been previously considered, but here's my thought:

1. Replace `test_module` in the matrix with a `shards` list:

    - shards:
      - {"root": "cms", "shard": "0/2"}
      - {"root": "cms", "shard": "1/2"}
      - {"root": "*", "shard": "0/10"}
      - {"root": "*", "shard": "1/10"}
      - {"root": "*", "shard": "2/10"}
      - {"root": "*", "shard": "3/10"}
      - {"root": "*", "shard": "4/10"}
      - {"root": "*", "shard": "5/10"}
      - {"root": "*", "shard": "6/10"}
      - {"root": "*", "shard": "7/10"}
      - {"root": "*", "shard": "8/10"}
      - {"root": "*", "shard": "9/10"}
2. Add a step that builds a shard:
    1. Make a list of modules. I don't have an exact algorithm here, but probably we can use pytest to collect them.
    2. If `matrix.shard.root == 'cms'` only keep module paths starting with `cms/`, otherwise keep only the module paths that don't start with `cms/`.
    3. Filter the shards: Concatenate each path with the PR number (as a weak source of randomness) and run SHA1. Get the hash bytes and modulo the second number in the `matrix.shards.shard`, then check if the result is equal to the first number. For example, the SHA-1 of `lms/djangoapps/verify_student/tests/test_tasks.py|12345` is `3f0a540d88a2c94da51d34d5ab920b4bdabca8f2`. `0x...a8f2 % 10 = 4`. So for shard `4/10`, filter down the list of test modules to just those with hash modulo `4`.
3. Run tests on just the resulting filtered list.

Reply:
Haha, sounds like I was well on my way to reinventing pytest-split! :-P But the later discussion allays my concerns about manual sharding. Only remaining concern here is that there should be a comment explaining how to do the shard balancing and linking to an ADR.